### PR TITLE
Fix `dpnp.tensor.acosh` handling of complex(+-0, NaN)

### PR DESCRIPTION
**Pull request by vlad-perevezentsev**
This PR proposes to fix issue #2880  where `dpnp.tensor.acosh(complex(0, NaN))` returned `NaN + NaN j` instead of `NaN ± π/2 j` as required by the Array API specification

The issue was caused by an incorrect early-return branch in acosh.hpp:
```
if (std::isnan(ry)) {
    return resT{ry, ry};
}
```
This special case bypassed the general logic which already produces the correct result.

The fix removes this unnecessary branch making the behavior consistent with both the Array API specification and NumPy
```
In [1]: import dpnp.tensor as dpt

In [2]: import numpy as np

In [3]: a = dpt.asarray([complex(+0, np.nan)])

In [4]: dpt.acosh(a)
Out[4]: usm_ndarray([nan+1.57079633j])

In [5]: a = np.asarray([complex(+0, np.nan)])

In [6]: np.acosh(a)
Out[6]: array([nan+1.57079633j])
```

Also adds a new `test_acosh_zero_nan` test to cover this case

- [X] Have you provided a meaningful PR description?
- [X] Have you added a test, reproducer or referred to an issue with a reproducer?
- [X] Have you tested your changes locally for CPU and GPU devices?
- [X] Have you made sure that new changes do not introduce compiler warnings?
- [X] Have you checked performance impact of proposed changes?
- [ ] Have you added documentation for your changes, if necessary?
- [X] Have you added your changes to the changelog?

### CODE REVIEW

File: dpnp/tests/tensor/elementwise/test_hyperbolic.py [hunk 1]
```diff
@@ -200,3 +200,24 @@ def test_hyper_real_special_cases(np_call, dpt_call, dtype):
 
     tol = 8 * dpt.finfo(dtype).resolution
     assert_allclose(dpt.asnumpy(dpt_call(yf)), Y_np, atol=tol, rtol=tol)
+
+
+@pytest.mark.parametrize("dtype", ["c8", "c16"])
+def test_acosh_zero_nan(dtype):
```
Comment (antonwolfy) [inline dpnp/tests/tensor/elementwise/test_hyperbolic.py:208]: new test is failing

File: dpnp/tests/tensor/elementwise/test_hyperbolic.py [hunk 2]
```diff
@@ -200,3 +200,24 @@ def test_hyper_real_special_cases(np_call, dpt_call, dtype):
 
     tol = 8 * dpt.finfo(dtype).resolution
     assert_allclose(dpt.asnumpy(dpt_call(yf)), Y_np, atol=tol, rtol=tol)
+
+
+@pytest.mark.parametrize("dtype", ["c8", "c16"])
```
Comment (antonwolfy) [inline dpnp/tests/tensor/elementwise/test_hyperbolic.py]: Please use `_complex_fp_dtypes` here

File: dpnp/tests/tensor/elementwise/test_hyperbolic.py [hunk 3]
```diff
@@ -200,3 +200,24 @@ def test_hyper_real_special_cases(np_call, dpt_call, dtype):
...
+def test_acosh_zero_nan(dtype):
+    # check acosh(0 + NaN j) = NaN ± πj/2
```
Comment (antonwolfy) [inline dpnp/tests/tensor/elementwise/test_hyperbolic.py]: I guess to be consistent with left part:
```suggestion
    # check acosh(0 + NaN j) = NaN ± π/2 j
```

File: dpnp/tests/tensor/elementwise/test_hyperbolic.py [hunk 4]
```diff
@@ -200,3 +200,24 @@ def test_hyper_real_special_cases(np_call, dpt_call, dtype):
...
+    q = get_queue_or_skip()
+    skip_if_dtype_not_supported(dtype, q)
+
+    x = [complex(+0.0, np.nan), complex(-0.0, np.nan)]
+
+    xf = np.array(x, dtype=dtype)
+    yf = dpt.asarray(xf, dtype=dtype, sycl_queue=q)
+
+    with np.errstate(all="ignore"):
+        Y_np = np.arccosh(xf)
```
Comment (antonwolfy) [inline dpnp/tests/tensor/elementwise/test_hyperbolic.py]: NumPy also has `acosh` alias:
```suggestion
        Y_np = np.acosh(xf)
```

File: dpnp/tests/tensor/elementwise/test_hyperbolic.py [hunk 5]
```diff
@@ -200,3 +200,24 @@ def test_hyper_real_special_cases(np_call, dpt_call, dtype):
...
+
+    Y_dpt = dpt.asnumpy(dpt.acosh(yf))
+
+    for i in range(len(x)):
```
Comment (antonwolfy) [inline dpnp/tests/tensor/elementwise/test_hyperbolic.py]: we need to place a comment that NumPy returns wrong result: NaN + NaN j
and probably to add a link to the issue in NumPy repo

File: dpnp/tests/tensor/elementwise/test_hyperbolic.py [hunk 6]
```diff
@@ -200,3 +200,24 @@ def test_hyper_real_special_cases(np_call, dpt_call, dtype):
...
+        assert np.isnan(Y_np[i].real) == np.isnan(Y_dpt[i].real)
```
Comment (antonwolfy) [inline dpnp/tests/tensor/elementwise/test_hyperbolic.py]: why do we need a loop? isn't the same as like:
```python
assert np.isnan(Y_np.real) == np.isnan(Y_dpt.real)
```

File: dpnp/tests/tensor/elementwise/test_hyperbolic.py (hunk 5 above)
Comment (ndgrigorian) [reply dpnp/tests/tensor/elementwise/test_hyperbolic.py]: conda-forge numpy gives the `nan, nan` result, while wheels give the `0, nan` result, because of something to do with numpy dispatching

https://github.com/numpy/numpy/issues/29167#issuecomment-2968099392

probably, it would be best to just not use numpy for testing these kinds of edge case results

Comment (vlad-perevezentsev) [reply dpnp/tests/tensor/elementwise/test_hyperbolic.py]: Yes, I have noticed that too
It is better not to compare it to NumPy at all
Thank you!

File: dpnp/tests/tensor/elementwise/test_hyperbolic.py (hunk 6 above)
Comment (vlad-perevezentsev) [reply dpnp/tests/tensor/elementwise/test_hyperbolic.py]: Done
Thank you!

File: dpnp/tests/tensor/elementwise/test_hyperbolic.py (hunk 1 above)
Comment (vlad-perevezentsev) [reply dpnp/tests/tensor/elementwise/test_hyperbolic.py:208]: I have updated this test

File: dpnp/tests/tensor/elementwise/test_hyperbolic.py (hunk 5 above)
Comment (antonwolfy) [reply dpnp/tests/tensor/elementwise/test_hyperbolic.py]: But both cases seem invalid, we expect `NaN ± π/2 j` based on the spec. Should we report then?

Comment (ndgrigorian) [reply dpnp/tests/tensor/elementwise/test_hyperbolic.py]: > But both cases seem invalid, we expect `NaN ± π/2 j` based on the spec. Should we report then?

PyPI isn't incorrect
```
In [1]: import numpy as np

In [2]: vals = [complex(np.nan, -np.inf), complex(-np.inf, np.nan), complex(0, np.nan)]

In [3]: x = np.array(vals)

In [4]: np.arcsinh(x[0]), np.arcsin(x[1]), np.arccosh(x[2])
Out[4]:
(np.complex128(inf+nanj),
 np.complex128(nan+infj),
 np.complex128(nan+1.5707963267948966j))
```
the last case is `NaN + pi/2`, it's only conda-forge which is incorrect.

I don't think we shouldn't file, just noting that, they are transiently aware of it and it's platform-dependent for NumPy failure

File: dpnp/tests/tensor/elementwise/test_hyperbolic.py [hunk 7]
```diff
@@ -200,3 +202,21 @@ def test_hyper_real_special_cases(np_call, dpt_call, dtype):
...
+    x = [complex(+0.0, np.nan), complex(-0.0, np.nan)]
+
+    xf = np.array(x, dtype=dtype)
+    yf = dpt.asarray(xf, dtype=dtype, sycl_queue=q)
+
+    Y_dpt = dpt.asnumpy(dpt.acosh(yf))
+
+    pi_half = np.full(len(x), np.pi / 2, dtype=xf.real.dtype)
+    assert np.all(np.isnan(Y_dpt.real))
+    assert_allclose(abs(Y_dpt.imag), pi_half, atol=1e-6)
```
Comment (antonwolfy) [inline dpnp/tests/tensor/elementwise/test_hyperbolic.py]: I guess it might be simplified:
```suggestion
    assert np.isnan(Y_dpt.real).all()
    assert_allclose(np.abs(Y_dpt.imag), np.pi / 2, atol=1e-6)
```

File: dpnp/tests/tensor/elementwise/test_hyperbolic.py (hunk 5 above)
Comment (antonwolfy) [reply dpnp/tests/tensor/elementwise/test_hyperbolic.py]: I see, I thought the same faulty result for PyPI.

**Review by antonwolfy — APPROVED** [review]
The only minor nit in the tests. The remaining is absolutely LGTM!
Thank you @vlad-perevezentsev

File: dpnp/tests/tensor/elementwise/test_hyperbolic.py (hunk 7 above)
Comment (vlad-perevezentsev) [reply dpnp/tests/tensor/elementwise/test_hyperbolic.py]: `assert_allclose` in dpnp uses `strict=True` thich requires matching shapes so I use `pi_half` as an array
```
E       AssertionError: 
E       Not equal to tolerance rtol=1e-07, atol=1e-06
E       
E       (shapes (2,), () mismatch)
E        ACTUAL: array([1.570796, 1.570796])
E        DESIRED: array(1.570796)

dpnp/tests/testing/array.py:75: AssertionError
```

Comment (vlad-perevezentsev) [reply dpnp/tests/tensor/elementwise/test_hyperbolic.py]: I have set `strict=False` for that

**Review by antonwolfy — APPROVED** [review]
